Preserve source location information during parsing

neuroevolutus opened the pull request:
I have implemented the functionality necessary to preserve source information through the parsing stage, but I have not added any tests yet. I was looking to see if you had any guidance about the number and kind of tests we would like to see. I was thinking of modifying the parser tests so that the provided streams of tokens used for input would also have artificially annotated `source_extent` values, but I understand that may blow up the size of the source code of the tests if we were to check for equality on source location values for all such parser tests. I would be truly grateful for your input.

neuroevolutus committed:
Preserve source location information during parsing

tekknolagi commented:
This looks pretty good! Good bones. I think:

* We don't need to worry about testing this /too thoroughly/ because it's a nice-to-have. So a couple of spot checks on some AST nodes, some tests for the coalescing, etc should be good.
* If there is an API trying to "get out" then we should figure out what it looks like to clean up the parser a bit. Should we be keeping track of source extents better in the parser? Do we need an auxiliary function to construct nodes? etc
* I don't know that we need to drop `frozen=True` because using `object.__setattr__` works OK on frozen nodes (and we've disabled the automatic comparison bits for the `source_extent` attr)

Basically I don't have any strong recommendations right now other than don't worry too much about testing. Excited to see what you come up with

tekknolagi commented on scrapscript.py:
```diff
@@ -521,68 +538,92 @@ def parse_unary(tokens: Peekable, p: float) -> "Object":
         else:
             l = parse(tokens)
-        next(tokens)
+            next(tokens)
```
This worries me. Why are we indenting `next`? Why did this not affect tests?

neuroevolutus replied:
Thanks for pointing that out. `next` is indented because we are now also invoking `next` in the `if` branch a few lines above. As to why I decided against just calling `next` at the end of the entire `if`-`else` clause, I was thinking perhaps it might not be desirable to include the parentheses around a parenthesized expression as part of its source extent. That is to say, I was thinking a nested expression like `(((((1 + 2)))))` should only track the innermost expression `1 + 2` as part of its source extent. That meant that I had to call `next` in the `if` branch so I could get the source extent of the closing parenthesis of a `Hole`. However, if it's desirable to include the enclosing parentheses as part of the source extent of an expression, I would be happy to change this.

tekknolagi replied:
Oh, huh. Tricky. I think what you have (no parens) is the move. Nice

tekknolagi commented:
This is totally out of scope for this change but for later: we should make good use of this information! I can't find any well-known spanned error message libraries like Rust's ariadne. Maybe we should make a tiny one

neuroevolutus committed:
Add `source_equals` method to support testing of AST source extents

neuroevolutus committed:
Add TODO note for Max

neuroevolutus committed:
Add tests to check for preservation of source extents during parsing

neuroevolutus committed:
Add function `make_source_annotated_object`

This function makes it easier to create AST objects that also have
`SourceExtent`s.

neuroevolutus committed:
Simplify `scrapscript.py` using `make_source_annotated_object`

neuroevolutus committed:
Simplify `scrapscript_tests.py` using `make_source_annotated_object`

neuroevolutus commented:
> This is totally out of scope for this change but for later: we should make good use of this information! I can't find any well-known spanned error message libraries like Rust's ariadne. Maybe we should make a tiny one

That would be awesome! I'll definitely have to take a look at `ariadne`.

neuroevolutus committed:
Fix errors from `ruff check`

neuroevolutus commented:
> If there is an API trying to "get out" then we should figure out what it looks like to clean up the parser a bit. Should we be keeping track of source extents better in the parser? Do we need an auxiliary function to construct nodes? etc

Thanks so much for the suggestion, Max. After thinking about it, I realised I could create a `make_source_annotated_object` function by cribbing your idea from the `make_token` function. Hopefully, this makes the code a bit more declarative.

neuroevolutus commented:
> * I don't know that we need to drop `frozen=True` because using `object.__setattr__` works OK on frozen nodes (and we've disabled the automatic comparison bits for the `source_extent` attr)

I might be doing something wrong, but I still seem to get `FrozenInstanceError`s when trying to reintroduce `frozen=True` and using `Object.__setattr__`.

tekknolagi commented:
```python
import dataclasses

@dataclasses.dataclass(frozen=True)
class C:
    x: int

c = C(3)
print(c)
object.__setattr__(c, "x", 4)
print(c)
```

I will look into this change more in detail tonight or tomorrow. Overall: wonderful!

I worry about including a bunch of new `source_equals` functions that are only used in testing. I think we can test this "level by level" where we test it for an int in one test, then a binop in another, then ... and assume the subtrees are well-constructed. That way we don't need to have a bunch of otherwise unused recursive functions hanging around.

Also: we should do some end-to-end tests where we go from text to testing source extents on ASTs.

tekknolagi commented on scrapscript.py:
```diff
@@ -591,19 +637,23 @@ def parse_unary(tokens: Peekable, p: float) -> "Object":
-            return Float(-r.value)
-        return Binop(BinopKind.SUB, Int(0), r)
+            return make_source_annotated_object(Float, source_extent, -r.value)
+        binop = Binop(BinopKind.SUB, Int(0), r)
```
`make_source_annotated_object` here?

tekknolagi commented on scrapscript.py:
```diff
@@ -617,7 +667,10 @@ def parse_binary(tokens: Peekable, p: float) -> "Object":
-            l = Apply(l, parse_binary(tokens, pr))
+            arg = parse_binary(tokens, pr)
+            new_l = Apply(l, arg)
+            new_l.source_extent = l.source_extent.coalesce(arg.source_extent) if l.source_extent else None
```
`make_source_annotated_object`?

tekknolagi commented on scrapscript.py:
```diff
@@ -627,31 +680,69 @@ def parse_binary(tokens: Peekable, p: float) -> "Object":
-            l = Assign(l, parse_binary(tokens, pr))
+            value = parse_binary(tokens, pr)
+            l = make_source_annotated_object(
+                Assign, l.source_extent.coalesce(value.source_extent) if l.source_extent else None, l, value
```
Do you think it's useful to have a free/global function to coalesce and do the conditional checking to clean things up?

neuroevolutus replied:
That's a good idea. I have created a new function called `join_source_extents` to address this.

neuroevolutus committed:
Add more invocations of `make_source_annotated_object`

neuroevolutus committed:
Use `make_source_annotated_object` in more tests

neuroevolutus committed:
Re-freeze AST objects

neuroevolutus commented:
> I will look into this change more in detail tonight or tomorrow.

This is my bad. I had just missed a couple of places in which I needed to use `object.__setattr__`.

neuroevolutus committed:
Add function `join_source_extents`

neuroevolutus committed:
Simplify parsing logic using `join_source_extents`

neuroevolutus commented:
> I worry about including a bunch of new source_equals functions that are only used in testing. I think we can test this "level by level" where we test it for an int in one test, then a binop in another, then ... and assume the subtrees are well-constructed. That way we don't need to have a bunch of otherwise unused recursive functions hanging around.

I see what you mean. I agree that having all these dubiously necessary `source_equals` functions are best avoided.

> Also: we should do some end-to-end tests where we go from text to testing source extents on ASTs.

What would you think about having potentially fewer of the above-mentioned incremental unit tests if we have a(n) end-to-end test(s) that exercises all the most important combinations of AST nodes?

tekknolagi commented:
Yeah for this I think we need only a couple incremental tests (one or two atomic types, one or two compound types) and then some end-to-end. Again, low stress here

neuroevolutus committed:
Revert "Add `source_equals` method to support testing of AST source extents"

This reverts commit 1605fd9fb978ae64ae2ea43378b19ca59f6a8ea0.

tekknolagi commented:
Just checking in - are you waiting for something from me on this PR?

neuroevolutus committed:
Include parentheses in source extents of expressions

neuroevolutus committed:
Fix tests

neuroevolutus committed:
Add source extent test for parenthesized expression

neuroevolutus committed:
Add a couple of end-to-end tests for source extents

neuroevolutus commented:
> Just checking in - are you waiting for something from me on this PR?

Oh, no, not at all. Thanks for checking in with me, Max. I'm almost ready to get your review again.

neuroevolutus committed:
Include hash symbol in variant source extent

neuroevolutus committed:
Apply eta reduction

neuroevolutus committed:
Format `scrapscript_tests.py`

neuroevolutus committed:
Change duplicate test name

neuroevolutus committed:
Remove unused variable

neuroevolutus committed:
Fix mypy errors

neuroevolutus committed:
Add source extent to int literal in test

neuroevolutus committed:
Rename local variables in test

neuroevolutus commented:
As a small note, I decided to undo my decision to exclude parentheses from the source extents of parenthesized expressions. I realized that 1) it would complicate maintaining the source extents of larger AST nodes that contained parenthesized sub-expressions (since it would cause them by default to not include the closing parentheses of a terminal parenthesized sub-expression) and 2) that people won't heavily nest parentheses in practice just to make life harder for themselves.

In a similar vein, I also undid my decision to exclude the pound sign from the source extents of variant expressions.

tekknolagi commented on scrapscript_tests.py:
```diff
@@ -1199,6 +1200,175 @@ def test_apply_with_variant_args(self) -> None:
+            start=SourceLocation(lineno=1, colno=1, byteno=0), end=SourceLocation(lineno=1, colno=1, byteno=0)
+        )
+        int_lit.source_extent = source_extent
+        int_ast = make_source_annotated_object(Int, source_extent, 1)
```
I don't think you need to make the AST nodes and can instead compare against the raw source_extent (here and elsewhere)

neuroevolutus replied:
Darn, you're right. That was totally redundant. Thanks for pointing that out.

tekknolagi commented on scrapscript.py:
```diff
@@ -521,68 +544,94 @@ def parse_unary(tokens: Peekable, p: float) -> "Object":
+        cases_source_extents = [case_branch.source_extent for case_branch in cases]
+        return make_source_annotated_object(
+            MatchFunction,
+            reduce(
```
Nit: we already have a loop and can join them in the loop instead of using `reduce`

tekknolagi reviewed (approved):
Love this. Great stuff. Left some minor comments for readability and cleanup

tekknolagi committed:
Merge branch 'trunk' into parser-line-column-byte-numbers

tekknolagi commented:
Small changes can be follow-ups as needed; wanted to get your changes in so we can move onto other exciting things like better error reporting in the parser, interpreter, etc
